pkg/agent/packetcapture/packetcapture_controller.go

```diff
@@ -0,0 +1,771 @@
+	}
+	return podIP, nil
+}
+
+func (c *Controller) parseIPs(pc *crdv1alpha1.PacketCapture) (srcIP, dstIP net.IP, err error) {
+	if pc.Spec.Source.Pod != nil {
+		srcIP, err = c.getPodIP(pc.Spec.Source.Pod)
+	} else if pc.Spec.Source.IP != nil {
+		srcIP = net.ParseIP(*pc.Spec.Source.IP)
+		if srcIP == nil {
+			err = fmt.Errorf("invalid source IP address: %s", *pc.Spec.Source.IP)
+		}
+	}
+
+	if pc.Spec.Destination.Pod != nil {
+		dstIP, err = c.getPodIP(pc.Spec.Destination.Pod)
+	} else if pc.Spec.Destination.IP != nil {
+		dstIP = net.ParseIP(*pc.Spec.Destination.IP)
+		if dstIP == nil {
+			err = fmt.Errorf("invalid destination IP address: %s", *pc.Spec.Destination.IP)
+		}
+	}
+	return
+}
+
+func (c *Controller) getUploaderByProtocol(protocol storageProtocolType) (sftp.Uploader, error) {
+	if protocol == sftpProtocol {
+		return c.sftpUploader, nil
+	}
+	return nil, fmt.Errorf("unsupported protocol %s", protocol)
+}
+
+func (c *Controller) generatePacketsPathForServer(name string) string {
+	return name + ".pcapng"
+}
+
+func (c *Controller) uploadPackets(pc *crdv1alpha1.PacketCapture, outputFile afero.File) error {
+	klog.V(2).InfoS("Uploading captured packets for PacketCapture", "name", pc.Name)
+	uploader, err := c.getUploaderByProtocol(sftpProtocol)
+	if err != nil {
+		return fmt.Errorf("failed to upload packets while getting uploader: %w", err)
+	}
+	if _, err := outputFile.Seek(0, 0); err != nil {
+		return fmt.Errorf("failed to upload to the file server while setting offset: %v", err)
+	}
+	authSecret := v1.SecretReference{
+		Name:      fileServerAuthSecretName,
+		Namespace: env.GetAntreaNamespace(),
+	}
+	serverAuth, err := auth.GetAuthConfigurationFromSecret(context.TODO(), auth.BasicAuthenticationType, &authSecret, c.kubeClient)
+	if err != nil {
+		klog.ErrorS(err, "Failed to get authentication for the file server", "name", pc.Name, "authSecret", authSecret)
+		return err
+	}
+	if serverAuth.BasicAuthentication == nil {
+		return fmt.Errorf("failed to get basic authentication info for the file server")
+	}
+	cfg := &ssh.ClientConfig{
+		User: serverAuth.BasicAuthentication.Username,
+		Auth: []ssh.AuthMethod{ssh.Password(serverAuth.BasicAuthentication.Password)},
+		// #nosec G106: skip host key check here and users can specify their own checks if needed
+		HostKeyCallback: ssh.InsecureIgnoreHostKey(),
+		Timeout:         time.Second,
+	}
+	return uploader.Upload(pc.Spec.FileServer.URL, c.generatePacketsPathForServer(pc.Name), cfg, outputFile)
+}
+
+func (c *Controller) updateStatus(name string, state *packetCaptureState) error {
+	toUpdate, getErr := c.packetCaptureLister.Get(name)
+	if getErr != nil {
+		klog.InfoS("Didn't find the original PacketCapture, skip updating status", "name", name)
+		return nil
+	}
+	conditions := []crdv1alpha1.PacketCaptureCondition{}
+	updatedStatus := crdv1alpha1.PacketCaptureStatus{
+		NumberCaptured: state.capturedPacketsNum,
+		FilePath:       state.filePath,
```

Comment:
there are data race to read them without lock.

Reply:
sounds good to me

Reply:
updated with shallow copied struct.